src/puffin/src/puffin_manager/cache_manager/moka_cache_manager.rs

```diff
@@ -0,0 +1,941 @@
+    ///   4. background routine removes the file or directory
+    delete_queue: Sender<DeleteTask>,
+}
+
+impl MokaCacheManager {
+    #[allow(unused)]
+    pub async fn new(base_dir: PathBuf, max_size: u64) -> Result<Self> {
+        let recycle_bin = Cache::builder()
+            .time_to_live(Duration::from_secs(60))
+            .build();
+
+        let recycle_bin_cloned = recycle_bin.clone();
+        let cache = Cache::builder()
+            .max_capacity(max_size)
+            .weigher(|_: &String, v: &CacheValue| v.weight())
+            .async_eviction_listener(move |k, v, _| {
+                let recycle_bin = recycle_bin_cloned.clone();
+                async move {
+                    recycle_bin.insert(k.as_str().to_string(), v).await;
+                }
+                .boxed()
+            })
+            .build();
+
+        let (delete_queue, rx) = tokio::sync::mpsc::channel(DELETE_QUEUE_SIZE);
+        common_runtime::bg_runtime().spawn(Self::delete_routine(rx));
+
+        let manager = Self {
+            cache,
+            base_dir,
+            delete_queue,
+            recycle_bin,
+        };
+
+        manager.recover().await?;
+
+        Ok(manager)
+    }
+}
+
+#[async_trait]
+impl CacheManager for MokaCacheManager {
+    type Blob = Arc<FsBlobGuard>;
+    type Dir = Arc<FsDirGuard>;
+
+    async fn get_blob<'a>(
+        &self,
+        puffin_file_name: &str,
+        key: &str,
+        init_fn: Box<dyn InitBlobFn + Send + Sync + 'a>,
+    ) -> Result<Self::Blob> {
+        let cache_key = Self::encode_cache_key(puffin_file_name, key);
+
+        let v = self
+            .cache
+            .try_get_with(cache_key.clone(), async {
+                if let Some(v) = self.recycle_bin.get(&cache_key).await {
+                    return Ok(v);
+                }
+
+                let file_name = format!("{}.{}", cache_key, uuid::Uuid::new_v4());
+                let path = self.base_dir.join(&file_name);
+
+                let size = Self::write_blob(&path, &init_fn).await?;
+
+                let guard = Arc::new(FsBlobGuard {
+                    path,
+                    delete_queue: self.delete_queue.clone(),
+                });
+                Ok(CacheValue::File { guard, size })
+            })
+            .await
+            .context(CacheGetSnafu)?;
+
+        match v {
+            CacheValue::File { guard, .. } => Ok(guard),
+            _ => unreachable!(),
+        }
+    }
+
+    async fn get_dir<'a>(
+        &self,
+        puffin_file_name: &str,
+        key: &str,
+        init_fn: Box<dyn InitDirFn + Send + Sync + 'a>,
+    ) -> Result<Self::Dir> {
+        let cache_key = Self::encode_cache_key(puffin_file_name, key);
+
+        let v = self
+            .cache
+            .try_get_with(cache_key.clone(), async {
+                if let Some(v) = self.recycle_bin.get(&cache_key).await {
+                    return Ok(v);
+                }
```

Comment:
is it necessary to remove hit entries from recycle bin?

Reply:
amazing catch, remove is right